Add comment replies

rudigiesler opened the pull request:
Replies to comments should be supported. You should be able to add a reply through the django admin when viewing the original comment.

Commit:
Add comment replies to django admin (first draft, tests incomplete)

Commit:
Add view for django admin comment replies

Commit:
Complete tests for django admin comment replies

Commit:
Template changes for comment replies

Commit:
Add tests for comment replies on comment list view

Commit:
Add test for ignored fields in comment reply

rudigiesler commented:
This is ready for review. Relies on https://github.com/praekelt/molo/issues/64 to be finished before the tests will pass.

hodgestar commented on molo/commenting/admin.py:
```diff
@@ -28,6 +45,16 @@ def reported_count(self, obj):
+        if obj.parent is None:
+            reply_url = reverse(
+                'admin:commenting_molocomment_reply', args=(obj.id,))
+            return obj.comment + ' <a href="%s">(reply)</a>' % (reply_url)
```
Bikeshed: I'm wondering whether the brackets should be outside of the `<a>` tag so that it renders like "([reply](#))" rather than "[(reply)](#)"? Perhaps it would be even better if it were a button? Does molo admin support bootstrap or similar?

hodgestar commented on molo/commenting/tests/test_admin.py:
```diff
@@ -0,0 +1,149 @@
+            i.get('name'): i.get('value') or ''
+            for i in html.form.find_all('input')
+        }
+        data.pop(None)
```
We could do:

``` python
data = {
    ...
    for i in html.form.find_all('input') if i.get('name')
}
```

and avoid the `.pop`.

hodgestar replied:
I think you missed this one?

hodgestar commented:
Just noting that currently it looks like an admin could alter the reply URL and reply to non-root comment. Questions:
* Is this something we need to disallow from a security prespective? (I'm guessing not?)
* Will having replies to replies break anything?

hodgestar commented:
@miltontony suggested putting the reply link in a separate column. I'm -0 on that because it seems more natural to have it next to the message content, but if it's the standard Molo way, I'm happy with that.

Another thought I had is how do we communicate to admins why some rows have the "reply" link and some don't?

hodgestar commented:
Left some comments, but looks good. :)

miltontony commented:
@rudigiesler instead of BeautifulSoup we often use Django's `self.assertContains(response, 'test reply text')` or even the non-elegant `self.assertContains(response, '<a href="#">some link</a>')`. Not the prettiest but it does avoid the extra dependency on BeautifulSoup.

I'm keen to try out pyQuery for more intense HTML tests, but I think for simple things `self.assertContains` is sufficient.

Commit:
Update molo.core version to fix urls scaffolding bug

Commit:
Cleanup on dictionary comprehension

Commit:
Change add reply to separate column with icon

rudigiesler commented:
@hodgestar I think I've addressed all of your comments. With respects to the security aspect, I can't really comment on it. It won't break anything, besides replies to replies not showing up on some templates (depending on how the templates are done, the ones in this repo will handle it fine). We could maybe add something to disallow it?

hodgestar commented:
I think you missed one instance of the dictionary comprehension clean up, otherwise looks good to me.

Would like @miltontony to just indicate he's okay with admins possibly being able to reply to replies.

Commit:
Neaten up other dictionary comprehension

miltontony commented:
I'm comfortable with this. I don't expect it to break anything since we control the depth to which we display comments.
:+1:

hodgestar commented:
:+1:
